### PR TITLE
light: move dropout handling and invalid data to the provider

### PR DESCRIPTION
**Pull request by cmwaters**
## Description

Builds on top of #5298

Addresses: #5186

### CODE REVIEW

File: types/validator_set.go [hunk 1]
```diff
@@ -986,17 +986,21 @@ func ValidatorSetFromProto(vp *tmproto.ValidatorSet) (*ValidatorSet, error) {
...
 // the exact same validator set that corresponds to it without changing the proposer priority or power
-func ValidatorSetFromExistingValidators(valz []*Validator) (*ValidatorSet, error) {
-	if len(valz) == 0 {
-		return nil, errors.New("no validators given")
+// if any of the validators fail validate basic then an empty set is returned.
+func ValidatorSetFromExistingValidators(valz []*Validator) *ValidatorSet {
+	for _, val := range valz {
+		err := val.ValidateBasic()
+		if err != nil {
+			return &ValidatorSet{}
```
Comment (erikgrinaker) [inline types/validator_set.go:994]: We shouldn't just silently discard errors, it makes debugging much harder.

Comment (cmwaters) [reply types/validator_set.go:994]: should I panic. Is it a developer fault if some / all of the validators are empty? or should I return an error?

Comment (erikgrinaker) [reply types/validator_set.go:994]: Return an error. The input to this function can be an external source.

Comment (cmwaters) [reply types/validator_set.go:994]: ok

**Review by erikgrinaker — APPROVED** [review]
LGTM

File: light/client.go [hunk 2]
```diff
@@ -1061,58 +1033,19 @@ func (c *Client) replacePrimaryProvider() error {
...
-	}
-
-	if expectedHeight > 0 && l.Height != expectedHeight {
-		return fmt.Errorf("height mismatch, got: %d, expected: %d", l.Height, expectedHeight)
+		c.logger.Debug("Error on light block request from primary", "error", err)
+		replaceErr := c.replacePrimaryProvider()
+		if replaceErr != nil {
+			return nil, fmt.Errorf("%v. Tried to replace primary but: %w", err.Error(), replaceErr)
+		}
+		// replace primary and request a light block again
```
Comment (melekes) [inline light/client.go:1045]: looks like comment should go above replacePrimaryProvider or split into 2 parts

File: types/validator_set.go [hunk 3]
```diff
@@ -986,9 +986,13 @@ func ValidatorSetFromProto(vp *tmproto.ValidatorSet) (*ValidatorSet, error) {
 
 // ValidatorSetFromExistingValidators takes an existing array of validators and rebuilds
 // the exact same validator set that corresponds to it without changing the proposer priority or power
+// if any of the validators fail validate basic then an empty set is returned.
 func ValidatorSetFromExistingValidators(valz []*Validator) (*ValidatorSet, error) {
-	if len(valz) == 0 {
-		return nil, errors.New("no validators given")
+	for _, val := range valz {
+		err := val.ValidateBasic()
```
Comment (melekes) [inline types/validator_set.go:992]: not sure this is the right place for validating input data

**Pull request merged**